enumerable_methods.rb

```diff
@@ -0,0 +1,212 @@
+
+  def my_each_with_index
+    return enum_for(:my_each_with_index) unless block_given?
+
+    arr = *self
+    i = 0
+    while i < arr.length
+      yield(arr[i], i)
+      i += 1
+    end
+    self
+  end
+
+  # select
+
+  def my_select
+    return enum_for(:my_select) unless block_given?
+
+    arr = *self
+    new_arr = []
+    arr.my_each do |element|
+      next unless yield(element)
+
+      new_arr.push(element)
+    end
+    new_arr
+  end
+
+  # my_all?
+
+  def my_all?(arg = nil)
+    condition = true
+    my_each do |element|
+      if arg.is_a?(Integer)
+        if element != arg
+          condition = false
+          break
+        end
+      end
+      if arg.is_a?(Class)
+        if element.is_a?(arg) == false
+          condition = false
+          break
+        end
+      elsif arg.is_a?(Regexp)
+        if element.match?(arg) == false
+          condition = false
+          break
+        end
+      else
+        if block_given?
+          if yield(element) == false || element.nil?
+            condition = false
+            break
+          end
+        else
+          if element == false || element.nil?
+            condition = false
+            break
+          end
+        end
+      end
+    end
+    condition
+  end
+
+  # my_any?
+
+  def my_any?(arg = nil)
+    condition = false
+    my_each do |element|
+      if arg.is_a?(Integer)
+        if element == arg
+          condition = true
+          break
+        end
+      elsif arg.is_a?(Class)
+        if element.is_a?(arg)
+          condition = true
+          break
+        end
+      elsif arg.is_a?(Regexp)
+        if element.match?(arg)
+          condition = true
+          break
+        end
+      elsif block_given?
+        if yield(element) || element.nil?
+          condition = true
+          break
+        end
+      elsif !block_given?
+        if element == false || element.nil?
+          condition = false
+          break
+        end
+      elsif arg.nil? && !block_given?
+        break if element == false
+      end
+    end
+    condition
+  end
+
+  # my_none?
+
+  def my_none?(arg = nil)
+    !my_any?(arg)
+  end
+
+  # count
+
+  def my_count(arg = nil)
+    arr = *self
+    if block_given? && arg.nil?
+      new_arr = []
+      arr.my_each do |element|
+        next unless yield(element)
+
+        new_arr.push(element)
+      end
+      new_arr.length
+    elsif (!block_given? && !arg.nil?) || (block_given? && !arg.nil?)
+      new_arr = []
+      arr.my_each do |element|
+        new_arr.push(element) if element == arg
+      end
+      new_arr.length
+    else
+      arr.length
+    end
+  end
+
+  # my_map
+
+  def my_map(my_proc = nil)
+    return enum_for(:my_map) unless block_given?
+
+    new_arr = []
+    if my_proc
+      my_each { |i| new_arr.push my_proc.call(i) }
+    else
+      my_each { |i| new_arr.push(yield(i)) }
+    end
+    new_arr
+  end
+
+  # my_inject
+
+  def my_inject(arg1 = 0, arg2 = nil)
```

Comment:
- When a block is given with an initial value, it combines all elements of enum by applying a binary operation, specified by a block.

  ![image](https://user-images.githubusercontent.com/57726348/94055970-438aa600-fdfb-11ea-9728-207cbe16c5ed.png)

- When a symbol is specified **without an initial value** combines each element of the collection by applying the symbol as a named method. Your method throws an error - `NoMethodError: undefined method zero?' for :+:Symbol`.

- When a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named **method:range**. Your method throws an error - `NoMethodError: undefined method zero?' for :+:Symbol`.